File: src/ts/examples/read-http.ts
/** Reading over HTTP range requests -- the point of the format.
 *
 *      node examples/read-http.ts <url> [minX minY maxX maxY]
 *
 *  Opening reads only the header. A bbox query then reads a few index
 *  pages plus the matching features, so a query against a file of tens
 *  of gigabytes transfers kilobytes. The request count is printed
 *  throughout because that, not the wall clock, is what the format is
 *  designed to minimise.
 *
 *  `fromUrl` wraps the source in a BufferedRangeReader for you (unlike
 *  `fromReader`, which takes the source exactly as given) -- see
 *  custom-reader.ts for the difference that makes.
 *
 *  ALWAYS pass a bbox on a large file: with no bbox this would scan the
 *  whole thing.
 */
import { FcbReader, toCityJSONFeature } from '@cityjson/flatcitybuf'

const [url, ...rest] = process.argv.slice(2)
if (url === undefined || (rest.length !== 0 && rest.length !== 4)) {
  console.log('usage: node examples/read-http.ts <url> [minX minY maxX maxY]')
  process.exit(2)
}

// Counting fetch: the reader takes any fetch-compatible function, which
// is also how you would add auth headers, retries or a cache.
let requests = 0
const countingFetch: typeof globalThis.fetch = (...args) => {
  requests += 1
  return globalThis.fetch(...args)
}

const t0 = Date.now()
const reader = await FcbReader.fromUrl(url, { fetch: countingFetch })
const info = reader.header.info
const secs = (since: number) => ((Date.now() - since) / 1000).toFixed(1)

console.log(`${info.featuresCount} features, CityJSON ${info.version}`)
console.log(`opened in ${requests} HTTP request(s), ${secs(t0)}s`)

if (rest.length !== 4) {
  console.log('no bbox given; not scanning the whole file -- pass one')
} else {
  const t1 = Date.now()
  const hits = await reader.select({
    spatial: { kind: 'bbox', value: rest.map(Number) as [number, number, number, number] },
  })
  // featuresCount is the TOTAL number of matches, known before iterating.
  console.log(
    `${hits.featuresCount} feature(s) in the query bbox, ` +
      `${requests} HTTP request(s), ${secs(t1)}s`,
  )

  const t2 = Date.now()
  let n = 0
  for await (const feature of hits) {
    toCityJSONFeature(feature, reader.header)
    n += 1
  }
  console.log(`decoded all ${n}, ${requests} HTTP request(s) total, +${secs(t2)}s`)
}
await reader.close()

File: src/py/examples/read_http.py
"""Reading over HTTP range requests -- the point of the format.

    python examples/read_http.py <url> [minx miny maxx maxy]

Opening reads only the header. A bbox query then reads a few index
pages plus the matching features, so a query against a file of tens of
gigabytes transfers kilobytes. `request_count` is printed throughout
because the request count, not the wall clock, is what the format is
designed to minimise.

ALWAYS pass a bbox on a large file: with no bbox this scans everything.
"""

from __future__ import annotations

import sys
import time

import flatcitybuf as fcb


def main(argv: list[str]) -> int:
    url = argv[0]
    box = tuple(float(v) for v in argv[1:5]) if len(argv) == 5 else None

    t0 = time.time()
    source = fcb.HttpRangeReader(url)
    reader = fcb.FcbReader.open(source)
    info = reader.header.info
    print(f"{info.features_count} features, CityJSON {info.cityjson_version}")
    print(
        f"opened in {source.request_count} HTTP request(s), "
        f"{time.time() - t0:.1f}s"
    )

    if box is None:
        print("no bbox given; not scanning the whole file -- pass one")
        return 0

    t1 = time.time()
    hits = fcb.search_rtree(
        reader.range_reader,
        reader.header.layout.rtree_begin,
        info.features_count,
        info.index_node_size,
        box,  # type: ignore[arg-type]
    )
    print(
        f"{len(hits)} feature(s) in the query bbox, "
        f"{source.request_count} HTTP request(s), {time.time() - t1:.1f}s"
    )

    # feature_at reads through a window shared across calls, so walking
    # a sorted hit list costs a fetch per window, not one per feature.
    t2 = time.time()
    for hit in hits:
        fcb.to_cityjson_feature(reader.feature_at(hit), reader.header)
    print(
        f"decoded all {len(hits)}, {source.request_count} HTTP request(s) "
        f"total, +{time.time() - t2:.1f}s"
    )
    return 0


if __name__ == "__main__":
    if len(sys.argv) not in (2, 6):
        print(__doc__)
        raise SystemExit(2)
    raise SystemExit(main(sys.argv[1:]))
